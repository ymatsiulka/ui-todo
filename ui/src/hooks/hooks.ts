import { useDispatch, useSelector, useStore, TypedUseSelectorHook } from 'react-redux';
import { AppDispatch, RootState, AppStore } from 'store/store';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppStore: () => AppStore = useStore;
