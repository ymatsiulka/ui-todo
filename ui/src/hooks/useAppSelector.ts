import { useSelector, type TypedUseSelectorHook } from 'react-redux';
import { type RootState } from 'store/store';

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
