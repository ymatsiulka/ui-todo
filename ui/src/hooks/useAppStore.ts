import { useStore } from 'react-redux';
import { type AppStore } from 'store/store';

export const useAppStore: () => AppStore = useStore;
