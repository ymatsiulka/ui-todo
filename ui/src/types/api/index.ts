export * from './TodoResponse';
export * from './TodoRequest';
