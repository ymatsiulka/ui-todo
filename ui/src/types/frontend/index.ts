export * from './props/SpacerProps';
export * from './props/ButtonProps';
export * from './props/TodoPageStatusesProps';
