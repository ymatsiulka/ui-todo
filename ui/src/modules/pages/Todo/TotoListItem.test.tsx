import React, { useState } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { TodoPageStatuses } from 'types/frontend';
import TodoListItem from './TotoListItem';
import { useAppDispatch } from 'hooks';

const mockUseState = useState as jest.Mock;
const mockUseAppDispatch = useAppDispatch as jest.Mock;
const mockSetState = jest.fn();

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

jest.mock('hooks', () => ({
  useAppDispatch: jest.fn(),
}));

jest.mock('store/features/todos/todosSlice', () => ({
  checkTodo: jest.fn(),
  deleteTodo: jest.fn(),
}));

jest.mock('modules/shared', () => ({
  Button: jest.fn((props) => <button data-testid="button" {...props} />),
  Checkbox: jest.fn(({ isChecked, onCheckedHandler, ...props }) => (
    <input data-testid="checkbox" type="checkbox" checked={isChecked} onChange={onCheckedHandler} {...props} />
  )),
}));

beforeEach(() => {
  mockSetState.mockClear();
  mockUseState.mockImplementationOnce(() => [true, mockSetState]);
});

test('TodoListItem should render properly', async () => {
  // Arrange
  const mockOnDragStartHandler = jest.fn();
  const mockOnDragEnterHandler = jest.fn();
  const mockOnDragEndHandler = jest.fn();

  // Act
  const { asFragment, container } = render(
    <TodoListItem
      id={1}
      name="name"
      isCompleted={true}
      todoPageStatus={TodoPageStatuses.Completed}
      onDragStartHandler={mockOnDragStartHandler}
      onDragEnterHandler={mockOnDragEnterHandler}
      onDragEndHandler={mockOnDragEndHandler}
    />,
  );

  const listItem = container.firstChild;
  fireEvent.dragStart(listItem);
  fireEvent.dragEnter(listItem);
  fireEvent.dragEnd(listItem);

  // Assert
  expect(asFragment()).toMatchSnapshot();
  expect(mockOnDragStartHandler).toHaveBeenCalled();
  expect(mockOnDragEnterHandler).toHaveBeenCalled();
  expect(mockOnDragEndHandler).toHaveBeenCalled();
});

test('TodoListItem when hover should update hovering state', async () => {
  // Arrange
  // Act
  const { container } = render(
    <TodoListItem
      id={1}
      name="name"
      isCompleted={true}
      todoPageStatus={TodoPageStatuses.Completed}
      onDragStartHandler={jest.fn()}
      onDragEnterHandler={jest.fn()}
      onDragEndHandler={jest.fn()}
    />,
  );

  const listItem = container.firstChild;
  fireEvent.mouseOver(listItem);
  fireEvent.mouseOut(listItem);

  // Assert
  expect(mockSetState).toHaveBeenCalledTimes(2);
  expect(mockSetState).toHaveBeenCalledWith(true);
  expect(mockSetState).toHaveBeenCalledWith(false);
});

test('TodoListItem when delete button clicked should delete todo', async () => {
  // Arrange
  const mockDispatch = jest.fn();
  mockUseAppDispatch.mockReturnValueOnce(mockDispatch);

  // Act
  const { getByTestId } = render(
    <TodoListItem
      id={1}
      name="name"
      isCompleted={true}
      todoPageStatus={TodoPageStatuses.Completed}
      onDragStartHandler={jest.fn()}
      onDragEnterHandler={jest.fn()}
      onDragEndHandler={jest.fn()}
    />,
  );

  const deleteButton = getByTestId('button');
  fireEvent.click(deleteButton);

  // Assert
  expect(mockDispatch).toHaveBeenCalledTimes(1);
});

test('TodoListItem when checkbox clicked should change todo state', async () => {
  // Arrange
  const mockDispatch = jest.fn();
  mockUseAppDispatch.mockReturnValueOnce(mockDispatch);

  // Act
  const { getByTestId } = render(
    <TodoListItem
      id={1}
      name="name"
      isCompleted={true}
      todoPageStatus={TodoPageStatuses.Completed}
      onDragStartHandler={jest.fn()}
      onDragEnterHandler={jest.fn()}
      onDragEndHandler={jest.fn()}
    />,
  );

  const checkbox = getByTestId('checkbox');
  fireEvent.click(checkbox);

  // Assert
  expect(mockDispatch).toHaveBeenCalledTimes(1);
  expect(mockSetState).toHaveBeenCalledTimes(1);
});
