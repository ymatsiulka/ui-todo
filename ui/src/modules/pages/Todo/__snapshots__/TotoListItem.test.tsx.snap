// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should render properly 1`] = `
<DocumentFragment>
  <li
    class="todoCheckboxItem"
    data-value="1"
    draggable="true"
    id="1"
  >
    <input
      checked=""
      data-testid="checkbox"
      id="1"
      label="name"
      name="1"
      type="checkbox"
    />
    <button
      class="delete-todo-button"
      data-testid="button"
    >
      <span />
    </button>
  </li>
</DocumentFragment>
`;
